### PR TITLE
feature/1390-connect-up-redis

### PR DESCRIPTION
**Pull request by jonhindle**
### Context

> Does this issue have a Trello card?
https://trello.com/c/dupoKPmL/1390-implement-redis-for-caching-drupal-requests-in-the-frontend

> If this is an issue, do we have steps to reproduce?
Most calls are not cached or cached inmemory.

### Intent

> What changes are introduced by this PR that correspond to the above card?
Redis setup and content cache rejig.

> Would this PR benefit from screenshots?
n/a

### Considerations

> Is there any additional information that would help when reviewing this PR?
n/a

> Are there any steps required when merging/deploying this PR?
Redis cache needs enabling/creating in the target environment.

### Checklist

- [ ] This PR contains **only** changes related to the above card
- [ ] Tests have been added/updated to cover the change
- [ ] Documentation has been updated where appropriate
- [ ] Tested in Development

### CODE REVIEW

File: server/repositories/cmsApi.js [hunk 1]
```diff
@@ -71,6 +76,18 @@ class CmsApi {
     });
   }
 
+  async getCache(query) {
+    const key = query.getKey?.() || null;
+    const expiry = query.getExpiry?.() || 8640;
+    let data = key ? await this.#cache.get(key) : null;
+    if (!data) {
+      data = (await this.get(query)) || [];
+      if (key && expiry && data) await this.#cache.set(key, data, expiry);
```
Comment (andrewrlee) [inline server/repositories/cmsApi.js:85]: key and expiry won't ever be null will they?

Comment (jonhindle) [reply server/repositories/cmsApi.js:85]: Key theoretically might be - but thats an error, so I probably could/should throw an error in that scenario instead 🤔

Comment (andrewrlee) [reply server/repositories/cmsApi.js:85]: but data would then be null?

File: server/utils/caching/redisClient.js [hunk 2]
```diff
@@ -45,6 +43,7 @@ class RedisCachingStrategy {
   }
 
   async get(key) {
+    logger.info(`Getting cache for '${key}'.`);
```
Comment (andrewrlee) [inline server/utils/caching/redisClient.js:46]: this is going generate a lot of logs, it shouldn't be necessary as you should be able to see the redis calls in app insights as `dependencies`

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+      - hmpps
+    container_name: hub-frontend-redis
+    environment:
+      - ALLOW_EMPTY_PASSWORD=yes
+    ports:
+      - '6379:6379'
+
+networks:
+  hmpps:
+
```
Comment (leonk) [inline docker-compose.yml:16]: We already have a docker-compose setup in the parent project. https://github.com/ministryofjustice/prisoner-content-hub/blob/main/docker-compose.yml (with Redis enabled for Drupal there).
I wonder if we should move this stuff into there?  That way there is only one docker compose file to maintain.  (I know that you guys don't use this yet, but maybe good if you started to? Also now that we have automated db backups for Drupal, we could probably incorporate that into the local build, as that was always the blocker before).

File: server/repositories/cmsQueries/categoryContentQuery.js [hunk 4]
```diff
@@ -26,6 +27,14 @@ class CategoryContentQuery {
     this.query = `${queryWithoutOffset}&${getPagination(page, limit)}`;
   }
 
+  getKey() {
+    getCmsCacheKey('categoryContent', this.establishmentName);
+  }
+
+  getExpiry() {
+    return 3600;
```
Comment (leonk) [inline server/repositories/cmsQueries/categoryContentQuery.js:35]: This is content on a category page?  I think caching for one hour will cause confusion (i.e. I added a piece of content, where is it?  Or I had to remove something, but it's still there).

The one to cache would be the sub_terms request.  As this just displays series and subcategories, and so won't need to be updated as regularly.  Also DCM's don't add new series/categories, only studio team do, and they can have a better understanding about the caching.

File: server/repositories/cmsApi.js (hunk 1 above)
Comment (jonhindle) [reply server/repositories/cmsApi.js]: I threw an error for missing keys.

File: server/repositories/cmsQueries/categoryContentQuery.js (hunk 4 above)
Comment (jonhindle) [reply server/repositories/cmsQueries/categoryContentQuery.js:35]: I think we go through each call and agree what caches and for how long.

**Review by BarryGeeMOJ — APPROVED** [review]
lgtm

**Pull request merged**